WalletWasabi.Gui/ProgramBase.cs

```diff
@@ -0,0 +1,193 @@
+	{
+		public Global? Global { get; private set; }
+
+		// This is only needed to pass CrashReporter to AppMainAsync otherwise it could be a local variable in Main().
+		private CrashReporter CrashReporter { get; } = new CrashReporter();
+
+		public TerminateService TerminateService { get; }
+
+		public ProgramBase()
+		{
+			TerminateService = new TerminateService(TerminateApplicationAsync);
+		}
+
+		public void Run(string[] args)
+		{
+			bool runGui = false;
+			Exception? appException = null;
+
+			try
+			{
+				Global = CreateGlobal();
+				Locator.CurrentMutable.RegisterConstant(Global);
+				Locator.CurrentMutable.RegisterConstant(CrashReporter);
+
+				SetPlatformBaseDirectory(Global.DataDir);
+				AppDomain.CurrentDomain.UnhandledException += CurrentDomain_UnhandledException;
+				TaskScheduler.UnobservedTaskException += TaskScheduler_UnobservedTaskException;
+
+				runGui = ProcessCliCommands(Global, args);
+
+				if (CrashReporter.IsReport)
+				{
+					StartCrashReporter(args);
+				}
+				else if (runGui)
+				{
+					Logger.LogSoftwareStarted("Wasabi GUI");
+					BuildAndRunAvaloniaApp(args);
+				}
+			}
+			catch (Exception ex)
+			{
+				appException = ex;
+			}
+
+			TerminateAppAndHandleException(appException, runGui);
+		}
+
+		public virtual void BuildAndRunAvaloniaApp(string[] args)
+		{
+			throw new NotImplementedException();
+		}
+
+		public virtual void SetPlatformBaseDirectory(string datadir)
+		{
+			throw new NotImplementedException();
+		}
+
+		private Global CreateGlobal()
+		{
+			string dataDir = EnvironmentHelpers.GetDataDir(Path.Combine("WalletWasabi", "Client"));
+			Directory.CreateDirectory(dataDir);
+			string torLogsFile = Path.Combine(dataDir, "TorLogs.txt");
+
+			var uiConfig = new UiConfig(Path.Combine(dataDir, "UiConfig.json"));
+			uiConfig.LoadOrCreateDefaultFile();
+			var config = new Config(Path.Combine(dataDir, "Config.json"));
+			config.LoadOrCreateDefaultFile();
+			config.CorrectMixUntilAnonymitySet();
+			var walletManager = new WalletManager(config.Network, new WalletDirectories(dataDir));
+
+			return new Global(dataDir, torLogsFile, config, uiConfig, walletManager);
+		}
+
+		private bool ProcessCliCommands(Global global, string[] args)
+		{
+			var daemon = new Daemon(global, TerminateService);
+			var interpreter = new CommandInterpreter(Console.Out, Console.Error);
+			var executionTask = interpreter.ExecuteCommandsAsync(
+				args,
+				new MixerCommand(daemon),
+				new PasswordFinderCommand(global.WalletManager),
+				new CrashReportCommand(CrashReporter));
+			return executionTask.GetAwaiter().GetResult();
+		}
+
+		public async void AppMainAsync(string[] _)
+		{
+			try
+			{
+				await InitAppMainAsync();
+			}
+			catch (Exception ex)
+			{
+				// There is no other way to stop the creation of the WasabiWindow, we have to exit the application here instead of return to Main.
+				TerminateAppAndHandleException(ex, true);
+			}
+		}
+
+		public virtual Task InitAppMainAsync()
```

Comment:
AppMainAsync and friends are specific to the GUI program.cs, and not the fluent (currently only there as a remnant left over.) consider moving these out of the base class.